Replicator/tests/ReplicatorCollectionSGTest.cc

```diff
@@ -1889,3 +1889,133 @@ TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Pull iTunes deltas from Collection
     C4Log("-------- %.3f sec with deltas, %.3f sec without; %.2fx speed",
           timeWithDelta, timeWithoutDelta, timeWithoutDelta/timeWithDelta);
 }
+
+TEST_CASE_METHOD(ReplicatorCollectionSGTest, "Give SGW random rev history and conflicts", "[.CBL-2694]") {
+    // The idea of this test is to exceed the RevTree history limit, so we can see what happens when we use the
```

Comment:
I am not sure what's the motivation of the changes to the two constants, 20 -> 50 and 20 -> 100. I restored that changes, and this test also passes.

Reply:
The test is intended to see the behaviour of conflict resolution when we are sending SG a history that contains randomly generated rev history. So the test just uses whatever the values are of `maxHistory` and `maxRevTreeDepth`, creates some conflicts in the rev history, and ensures that conflict resolution behaves as expected. 

Looking at the test now though, I do need to update it a bit since I changed the values from (100/20) to (100/50) since I first made the test.

Reply:
My wonder is actually the following. There are three parts of the in the PR.
1. Substance, it changed two constant, kDefaultMaxRevTreeDepth, kDefaultMaxHistory
2. Log the gap when the random revIDs are substituted in the history.
3. This test.
I tried to remove (1) and the test also passed. So, is the main object of PR is to add a test? Is there a thing that did not work without this PR?

Reply:
The test doesn't test something that's internally broken, the issue stems from SG receiving a rev history from us that has a gap. There were complaints from certain customers about this warning in SG. The test is simply to test that with the increased values, and thereby the usage of randomly generated revIDs, the conflict resolution will still behave as it does normally.

I discussed it with SG team, Jim and Jens and SG team considered whether they could just drop the warning from SG when they receive a gap. This is what Ben said to me:

> Adam replied saying this and then I pointed him to the litecore code generating non-deterministic revision IDs:
> > We wouldn't create fake conflicts if we had a deterministic approach for generating the intermediate digests (either clone the digest of the latest revision, or hash from parent rev ID).  Have we already reviewed whether the rev tree pruning algorithm (even for non-branching rev trees) is happy with gaps in generation numbers?

> It seems odd that the comment around that area of litecore code suggests its generating them in order to send them, but then never does

> It still feels like litecore should be filling that gap for us - couchdb (and therefore SG) formats revisions in a way that doesn't allow you to have gaps

So I decided the best course of action was to increase the length of history that we send to SG, so it is much harder to perform enough local mutations between syncs to end up sending a rev history with a gap to SG. That is why I propose to increase `ReplicatorTuning::kDefaultMaxHistory`.

The drawback with that, is any history in-between the revTree depth and maxHistory is randomly generated (within the function `TreeDocument::_getRevisionHistory`). As it is randomly generated, this could easily cause conflicts with other clients, and we want to cause as little conflicts as we can.

The current max revTree depth of 20 was set when we used to store the bodies of all revs in the tree, so each rev we store in the history was quite storage-expensive. As we now only store the bodies of the current rev and the last rev we synced with remote, additional revs in the revTree aren't too expensive (I'm assuming just however many bytes the revIDs are). 

Therefore, to reduce the usage of randomly generated revIDs, I want to increase `kDefaultMaxRevTreeDepth`, as it shouldn't be too storage-expensive and will significantly reduce the possibility of conflicts between clients' rev history.

Reply:
Additionally, the reason for logging a warning to the client when we had to use randomly generated revIDs is because if the client is making more than 50 mutations between syncs with SG, there is probably something going wrong.

Reply:
So, the fundamental problem is that the client voluntarily limit the max-history. The protocol allows the reply to include "maxHistory." If it is absent, the client would voluntarily limit it to a number, which I assume is not part of the protocol. Now, we are in a bad situation. On one hand, we limit the history to, say, 20 or 100, but the last revision that the remote knows is deeper than the max. On the other hand, we want to include the last synced revision in the history. 
Either we don't self-impose the maxHistory, or the protocol allows gaps and SG should handle it. Of course, if SG puts maxHistory in the reply, it knows the history may be truncated.

Reply:
`                if (hasRemoteAncestor(revID)) {
                    // Always write the common ancestor, making room if necessary:
                    if (revsWritten == maxRevs)
                        removeLast();
                    append(revID);
                    break;
                } else {`
^^^ from _getRevisionHistory